dev/PowerNotifications/PowerNotifications.h

```diff
@@ -0,0 +1,158 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#pragma once
+
+#include <mutex>
+#include <PowerManager.g.h>
+#include <PowerNotificationsPal.h>
+#include <wil/resource.h>
+#include <wil/result_macros.h>
+
+namespace winrt::Microsoft::ProjectReunion::implementation
+{
+    using PowerEventHandle = Windows::Foundation::EventHandler<Windows::Foundation::IInspectable>;
+    using EventToken = winrt::event_token ;
+
+    struct CompositeBatteryStats
+    {
+        int   oldBatteryChargePercent, batteryChargePercent;
+        winrt::Microsoft::ProjectReunion::BatteryStatus     oldBatteryStatus, batteryStatus;
+        winrt::Microsoft::ProjectReunion::PowerSupplyStatus oldPowerSupplyStatus, powerSupplyStatus;
+    };
+
+    struct PowerManager
+    {
+        PowerManager() = default;
+
+        using eventType = wil::object_without_destructor_on_shutdown<winrt::event<PowerEventHandle>>;
```

Comment:
Should we just use a normal event?
@brialmsft any suggestions?

Reply:
Yeah, this carried over from my first draft of the C++/WinRT implementation. I know how to set this up properly in WRL, but all of the documentation for C++/WinRT I could find didn't point toward a robust way to implement static events. Should have followed up instead of leaving it as-is.

If we don't leak, you still have COM unload issues (which I'm all too familiar with...): this DLL unloads, the event helper objects destruct, and they call Release on the registered delegates which were implemented in other potentially-already-unloaded DLLs (most notably dangerous in ExitProcess and final CoUninitialize scenarios). I think leaking is *less* risky than the naive option of letting the helper objects destruct: if we leak them, then we never signal the events, so nobody should be making unsafe calls into the delegates via our leaked pointers. Which is why I went with leak-on-unload at first.

With that said, the information you and @kennykerr provided should help us do this the right way, where we have a persistent factory object to hang semantically static state off of, and don't need to use the DLL's global/static state.

Reply:
The point of the static lifetime store is that the objects there are run down *before* COM starts tearing down DLLs. So when the event starts releasing handlers, all the handlers are still good (DLL unloading hasn't started yet).

Reply:
Right. I was talking specifically about the naive implementation without static lifetime, where we have no guarantee of singleton-instancing on the factory object, and thus have no place to store WinRT-static state other than the DLL's static/global storage. That's the scenario that has COM unload concerns if you don't leak the event helper objects.

I agree that using static lifetime to cache the factory and provide singleton instancing is much cleaner.